src/argparcel/magic.py

```diff
@@ -0,0 +1,184 @@
+    # to the user.
+    source = inspect.getsource(cls)
+    tree = ast.parse(source)
+
+    # Find the class definition node; if it doesn't exist something is fundamentally
+    # broken, so we are happy with an assert.
+    class_node = next(
+        (
+            node
+            for node in tree.body
+            if isinstance(node, ast.ClassDef) and node.name == cls.__name__
+        ),
+        None,
+    )
+    assert class_node is not None
+
+    result: dict[str, str] = {}
+
+    # Walk through the class body to collect field docstrings. We're looking for pairs
+    # of field definition followed by a docstring.
+    for node, next_node in itertools.pairwise(class_node.body):
+        if isinstance(node, ast.AnnAssign) and isinstance(node.target, ast.Name):
+            field_name = node.target.id
+
+            if (
+                isinstance(next_node, ast.Expr)
+                and isinstance(next_node.value, ast.Constant)
+                and isinstance(next_node.value.value, str)
+            ):
+                result[field_name] = next_node.value.value
+
+    return result
+
+
+def _is_type_checking_block(node: ast.AST) -> typing.TypeGuard[ast.If]:
+    """Returns True if the given AST node is an `if TYPE_CHECKING:` block."""
+    if not isinstance(node, ast.If):
+        return False
+
+    test = node.test
+    if isinstance(test, ast.Name):
+        return test.id == "TYPE_CHECKING"
+    if isinstance(test, ast.Attribute):
+        return (
+            isinstance(test.value, ast.Name)
+            and test.value.id == "typing"
+            and test.attr == "TYPE_CHECKING"
+        )
+    return False
+
+
+def _module_globals_including_type_checking(  # noqa: C901
+    module: types.ModuleType, *, allowed_extra_names: frozenset[str]
+) -> dict[str, typing.Any]:
+    """Get the global namespace for `module`... including imports in TYPE_CHECKING.
+
+    A few semantics:
+        - of any imports identified in TYPE_CHECKING blocks, only names that appear in
+        `allowed_extra_names` will actually be imported.
+        - only names that are _not_ already known in the normal runtime module namespace
+          will be imported
+        - the minim
+    """
+    source = inspect.getsource(module)
+    globalns = dict(vars(module))
+
+    tree = ast.parse(source)
+
+    for node in tree.body:
+        if not _is_type_checking_block(node):
+            continue
+
+        for statement in node.body:
+            if isinstance(statement, ast.Import):
+                for alias in statement.names:
+                    if alias.name not in allowed_extra_names:
+                        continue
+                    asname = alias.asname or alias.name
+                    if asname in globalns:
+                        # Do not overwrite names that have already been loaded into the
+                        # namespace.
+                        continue
+                    globalns[asname] = importlib.import_module(alias.name)
```

Comment:
Here, and below, `import_module` takes a `package` argument which must be provided if we're trying to do a _relative_ import.

Reply:
6905635